app/src/main/java/tech/relaycorp/cogrpc/server/TLSCertificateGenerator.kt

```diff
@@ -0,0 +1,100 @@
+package tech.relaycorp.cogrpc.server
+
+import org.bouncycastle.asn1.x500.X500Name
+import org.bouncycastle.asn1.x500.X500NameBuilder
+import org.bouncycastle.asn1.x500.style.BCStyle
+import org.bouncycastle.asn1.x509.Extension
+import org.bouncycastle.asn1.x509.GeneralName
+import org.bouncycastle.asn1.x509.GeneralNamesBuilder
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo
+import org.bouncycastle.cert.X509CertificateHolder
+import org.bouncycastle.cert.X509v3CertificateBuilder
+import org.bouncycastle.crypto.params.AsymmetricKeyParameter
+import org.bouncycastle.crypto.util.PrivateKeyFactory
+import org.bouncycastle.operator.ContentSigner
+import org.bouncycastle.operator.DefaultDigestAlgorithmIdentifierFinder
+import org.bouncycastle.operator.DefaultSignatureAlgorithmIdentifierFinder
+import org.bouncycastle.operator.bc.BcRSAContentSignerBuilder
+import org.bouncycastle.util.encoders.Base64
+import java.math.BigInteger
+import java.security.KeyPairGenerator
+import java.security.PrivateKey
+import java.security.SecureRandom
+import java.security.cert.CertificateException
+import java.time.ZoneOffset.UTC
+import java.time.ZonedDateTime
+import java.util.Date
+
+class TLSCertificateGenerator(
+    val privateKey: PrivateKey,
+    val certificateHolder: X509CertificateHolder
+) {
+    fun exportPrivateKey() = derToPem(this.privateKey.encoded, "PRIVATE KEY")
+
+    fun exportCertificate() = derToPem(this.certificateHolder.encoded, "CERTIFICATE")
+
+    private fun derToPem(valueDer: ByteArray, pemLabel: String): ByteArray {
+        val valueBase64 = Base64.toBase64String(valueDer)
+        val valuePem = "-----BEGIN $pemLabel-----\n${valueBase64}\n-----END $pemLabel-----"
+        return valuePem.toByteArray()
+    }
+
+    companion object {
+        private const val RSA_KEY_MODULUS = 2048
+        const val SUBJECT_IP_ADDRESS = "192.168.43.1"
```

Comment:
I was meaning to check this before, but it seems that we can't trust the IP to always be `192.168.43.1`. Most of the times it will be, but some more recent devices are picking an IP address randomly inside `192.168.43.X`.

Reply:
Indeed, I forgot to mention I discovered that yesterday: https://stackoverflow.com/questions/61615270/how-to-get-the-ip-address-of-the-default-gateway-reliably-on-android-5

From the testing I did, the certificate can have this IP address even if it's wrong, as long as the client disables hostname verification.

Reply:
BTW, I added a note to one of the tickets in the Android Gateway to look into this, but I think it's OK in the courier. Worst case scenario, we'd have to require users of the Android Gateway to use DHCP when connecting to the courier, which is what 99% of them would do anyway.